### DISCUSSION TITLE
How do I add a signature pad/field to the PDF?

### DISCUSSION THREAD

**Discussion by ChadReitsma** [dompdf/dompdf#3151] [Q&A] [answered]
Hey guys,
Is there a way to add a signature field so that when the user opens the .pdf they can sign it inside Acrobat?

**Comment by bsweeney** [comment, marked as answer]
There isn't a way to do this using HTML, but take a look at the discussion in #2155 and see if that gets you where you need.

Here's sample code from that pull request (I haven't tested this so can't confirm that it's functional with the current release):
```
$dompdf = new Dompdf\Dompdf();
$dompdf->loadHtml('<html><body><strong>PDF with AcroForm and Signature</strong></body></html>');
$dompdf->render();

$cpdf = $dompdf->getCanvas()->get_cpdf();

$cpdf->addForm(Dompdf\Cpdf::ACROFORM_SIG_SIGNATURESEXISTS);
$cpdf->setStrokeColor(\Dompdf\Css\Color::parse('#000000'));

$fieldSigId = $cpdf->addFormField(Dompdf\Cpdf::ACROFORM_FIELD_SIG, 'Signature1', 100, 470, 350, 490);
$cpdf->line(99, 469, 351, 469);

file_put_contents('test.pdf', $dompdf->output());
```

**Comment by ChadReitsma** [comment]
Thanks Brian!  That worked.

Now I'm running into a different issue whereby I need to find the coordinates of a div with a specific "signature" class, and place the signature box in that location.   I tried using your callbacks which I found in other discussions:

```
$GLOBALS['frame_data'] = array();
$dompdf = new DOMPDF();
$dompdf->setCallbacks(
      array(
        'myCallbacks' => array(
          'event' => 'end_frame', 'f' => function ($infos) {
            $frame = $infos['frame'];
            $GLOBALS['frame_data'][] = $frame->get_node()->getAttribute("class");
          }
        )
      )
    );
$dompdf->load_html($html);
$dompdf->render();

```

This results in 
Fatal error: Uncaught Error: Cannot use object of type Dompdf\FrameDecorator\Text as array  on **Line: 7  ($frame = $infos['frame'];**

Using PHP 8.2

What am I doing wrong?

**Reply by bsweeney** [reply]
Using 2.x then it sounds like. The [callback arguments](https://github.com/dompdf/dompdf/wiki/Usage#callbacks) changed with 2.0.0. Try this:

```
$frame_data = [];
$dompdf = new Dompdf();
$dompdf->setCallbacks([
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$frame_data) {
            $frame_data[] = $frame->get_node()->getAttribute("class");
        }
    ]
]);
$dompdf->load_html($html);
$dompdf->render();
```

**Comment by ChadReitsma** [comment]
Thanks Brian,

It's still giving me some errors - looks like getAttribute() isn't a valid method for DOMText:
Fatal error: Uncaught Error: Call to undefined method DOMText::getAttribute() 

I tried finding the frame based on it's textContent instead:

```
$frame_data = array();
$dompdf = new Dompdf\Dompdf();
$dompdf->setCallbacks([
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$frame_data) {
            if ($frame->get_node()->textContent == "^^signature^^") {
                $frame_data[] = $frame->get_padding_box();
            }
        }
    ]
]);

$dompdf->loadHtml('<html><body><strong>PDF with AcroForm and Signature</strong><div class="signature">^^signature^^</div></body></html>');
$dompdf->render();
print_r($frame_data);
```

But now $frame_data has two elements when there should only be one?? Kinda strange.

**Reply by bsweeney** [reply]
No I think that makes sense. textContent gets the content of a [node plus its descendents](https://www.php.net/manual/en/class.domnode.php#domnode.props.textcontent).

Looking for a unique identifier is the right way to do it, you just have to be more particular about how you do your check.

```
$node = $frame->get_node();
if (
    $node->hasAttributes()
    && ($class = $node->attributes->getNamedItem("class") !== null)
    && $class === "signature"
) {
    $frame_data[] = $frame->get_padding_box();
}
```

**Reply by ChadReitsma** [reply]
Ah, that makes sense. Thank-you again for the quick reply!  I'll give it a shot!

**Comment by ChadReitsma** [comment]
I was able to get the X,Y coordinates but had to make a change to \lib\Cpdf.php on line 4334:

```
$this->o_acroform($this->acroFormId, 'font',
          ['objNum' => $font[0]['objNum'], 'fontNum' => $font[0]['fontNum']]);
```

Throws an error because the $font array starts at 1 (not 0) 

```
print_r($font);
Array ( [1] => Array ( [objNum] => 9 [fontNum] => 2 ) ) 
```

**Reply by bsweeney** [reply]
I'll need to review that code to see why it's assuming index 0 there if that's not necessarily the case.
